### PR TITLE
Change Default Table Border Color

### PR DESCRIPTION
**Pull request by K-Kumar-01**
Changes color from 000000 to CCCCCC

### CODE REVIEW

File: src/helpers/xml-builder.js [hunk 1]
```diff
@@ -349,9 +397,9 @@ const cssBorderParser = (borderString) => {
       ].includes(token)
     ) {
       // Accepted OOXML Values for border style: http://officeopenxml.com/WPtableBorders.php
-      if (['dashed', 'dotted', 'double', 'inset', 'outset'].includes(token)) {
+      if (['dashed', 'dotted', 'double', 'inset', 'outset', 'none'].includes(token)) {
         stroke = token;
-      } else if (['hidden', 'none'].includes(token)) {
+      } else if (token === 'hidden') {
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: Remind me why we aren't treating none and hidden the same again?

File: src/helpers/xml-builder.js [hunk 2]
```diff
@@ -324,11 +331,52 @@ const fixupMargin = (marginString) => {
...
 };
 
-const cssBorderParser = (borderString) => {
-  const tokens = borderString.split(' ');
-  let size = 0;
+const borderStyleParser = (style) => {
   let stroke = 'single';
-  let color = '000000';
+  if (
+    [
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: We have a section here for style, and a section below with the other comment also parsing style.  "DRY" (Don't repeat yourself) and make this  a function

File: src/helpers/xml-builder.js [hunk 3]
```diff
@@ -324,11 +331,52 @@ const fixupMargin = (marginString) => {
...
+      'ridges',
+      'inset',
+      'outset',
+      'hidden',
+      'none',
+      'windowtext', // tinyMCE has this border property
+    ].includes(style)
+  ) {
+    // Accepted OOXML Values for border style: http://officeopenxml.com/WPtableBorders.php
+    if (['dashed', 'dotted', 'double', 'inset', 'outset', 'none'].includes(style)) {
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: DRY

Comment (nicolasiscoding) [reply src/helpers/xml-builder.js]: Remind me why we aren't treating none and hidden the same again?

File: src/helpers/xml-builder.js [hunk 4]
```diff
@@ -349,9 +397,9 @@ const cssBorderParser = (borderString) => {
       ].includes(token)
     ) {
       // Accepted OOXML Values for border style: http://officeopenxml.com/WPtableBorders.php
-      if (['dashed', 'dotted', 'double', 'inset', 'outset'].includes(token)) {
+      if (['dashed', 'dotted', 'double', 'inset', 'outset', 'none'].includes(token)) {
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: DRY

File: src/helpers/xml-builder.js [hunk 5]
```diff
@@ -1430,116 +1479,339 @@ const buildTableCellProperties = (attributes) => {
...
-        colors: {
-          ...setUpDirectionalBorderColor(borderColor),
-        },
-        strokes: {
-          ...setUpDirectionalBorderStroke(borderStroke),
-        },
-      };
-    }
-  }
+const fixupTableCellBorder = (vNode, attributes, tableBorderOptions = {}, rowIndexEquivalent) => {
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: trying to parse through this and its hard to follow without prior context, can you add comments throughout this section to clarify what is happening here?

Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: Can you also add comments to explain the params for this like rowIndexEquivalent?

File: src/helpers/xml-builder.js [hunk 6]
```diff
@@ -1430,116 +1479,339 @@ const buildTableCellProperties = (attributes) => {
...
+      attributes.tableCellBorder.top = borderSizeParser(tableCellStyles[tableCellStyle])
+    } else if (tableCellStyle === 'border-bottom-width') {
+      if (rowIndexEquivalentLast !== -1) continue
+      attributes.tableCellBorder.bottom = borderSizeParser(tableCellStyles[tableCellStyle])
+    }
   }
 };
 
-const buildTableCell = async (vNode, attributes, rowSpanMap, columnIndex, docxDocumentInstance) => {
+const buildTableCell = async (vNode, attributes, rowSpanMap, columnIndex, docxDocumentInstance, rowIndexEquivalent) => {
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: add comments around params

File: src/helpers/xml-builder.js [hunk 7]
```diff
@@ -1797,9 +2068,41 @@ const buildTableRowProperties = (attributes) => {
   return tableRowPropertiesFragment;
 };
 
-const buildTableRow = async (vNode, attributes, rowSpanMap, docxDocumentInstance) => {
+const buildTableRow = async (vNode, attributes, rowSpanMap, docxDocumentInstance, rowIndexEquivalent) => {
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js:2517]: explain params

File: src/helpers/xml-builder.js [hunk 8]
```diff
@@ -1430,116 +1479,339 @@ const buildTableCellProperties = (attributes) => {
...
-      colors: { ...attributes.tableCellBorder.colors, left: borderColor },
-      strokes: { ...attributes.tableCellBorder.strokes, left: borderStroke },
-    };
+  const tableCellStyleKeys = Object.keys(tableCellStyles);
+
+  const rowIndexEquivalentFirst = rowIndexEquivalent.indexOf('first')
+  const rowIndexEquivalentLast = rowIndexEquivalent.indexOf('last')
+  // here this means that 
+  if (rowIndexEquivalentFirst !== -1) {
+    const indexOfBorder = tableCellStyleKeys.lastIndexOf('border')
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js:1480]: why are we looking for indexes? Can you add comments and explain

File: src/helpers/xml-builder.js [hunk 9]
```diff
@@ -1430,116 +1479,339 @@ const buildTableCellProperties = (attributes) => {
...
+    const indexOfBorderTopWidth = tableCellStyleKeys.lastIndexOf('border-top-width')
+    const indexOfBorderTopStyle = tableCellStyleKeys.lastIndexOf('border-top-style')
+    const indexOfBorderTopColor = tableCellStyleKeys.lastIndexOf('border-top-color')
+    let indexes = []
+
+    if (indexOfBorderTopStyle !== -1 && tableCellStyles['border-top-style'] === 'hidden') {
+      attributes.tableCellBorder.top = 0
+      attributes.tableCellBorder.strokes.top = 'hidden'
+    } else {
+      if (indexOfBorder !== -1) {
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js:1499]: what is going on here?

File: src/helpers/xml-builder.js [hunk 10]
```diff
@@ -1430,116 +1479,339 @@ const buildTableCellProperties = (attributes) => {
...
+  // here this means that 
+  if (rowIndexEquivalentFirst !== -1) {
+    const indexOfBorder = tableCellStyleKeys.lastIndexOf('border')
+    const indexOfBorderTop = tableCellStyleKeys.lastIndexOf('border-top')
+    const indexOfBorderTopWidth = tableCellStyleKeys.lastIndexOf('border-top-width')
+    const indexOfBorderTopStyle = tableCellStyleKeys.lastIndexOf('border-top-style')
+    const indexOfBorderTopColor = tableCellStyleKeys.lastIndexOf('border-top-color')
+    let indexes = []
+
+    if (indexOfBorderTopStyle !== -1 && tableCellStyles['border-top-style'] === 'hidden') {
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js:1495]: Can you explain hidden vs none?

File: src/helpers/xml-builder.js [hunk 11]
```diff
@@ -1430,116 +1479,339 @@ const buildTableCellProperties = (attributes) => {
...
+  const rowIndexEquivalentFirst = rowIndexEquivalent.indexOf('first')
+  const rowIndexEquivalentLast = rowIndexEquivalent.indexOf('last')
+  // here this means that 
+  if (rowIndexEquivalentFirst !== -1) {
+    const indexOfBorder = tableCellStyleKeys.lastIndexOf('border')
+    const indexOfBorderTop = tableCellStyleKeys.lastIndexOf('border-top')
+    const indexOfBorderTopWidth = tableCellStyleKeys.lastIndexOf('border-top-width')
+    const indexOfBorderTopStyle = tableCellStyleKeys.lastIndexOf('border-top-style')
+    const indexOfBorderTopColor = tableCellStyleKeys.lastIndexOf('border-top-color')
+    let indexes = []
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js:1490]: What does this array do?

File: src/helpers/xml-builder.js [hunk 12]
```diff
@@ -1820,7 +2123,26 @@ const buildTableRow = async (vNode, attributes, rowSpanMap, docxDocumentInstance
       );
     }
     if (vNode.properties.style) {
-      fixupTableCellBorder(vNode, modifiedAttributes);
+      fixupTableCellBorder(vNode, modifiedAttributes, docxDocumentInstance.tableBorders, rowIndexEquivalent);
+    } else {
+      // it means that border style provided was none and border-attribute given was none.
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: Can you clarify this?

File: src/helpers/xml-builder.js [hunk 13]
```diff
@@ -2030,105 +2353,151 @@ const buildTable = async (vNode, attributes, docxDocumentInstance) => {
...
-          insideV: borderColor,
-        };
-      } else {
-        tableBorders.insideV = 0;
-        tableBorders.insideH = 0;
+    } else if (tableAttributes.border) {
+      modifiedAttributes.isTableBorderAttributeGiven = true
+
+      // here we have passed a string which is not 0. In such cases, the borderSize becomes 1
+      borderSize = 1
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: move this into consts

File: src/helpers/xml-builder.js [hunk 14]
```diff
@@ -2030,105 +2353,151 @@ const buildTable = async (vNode, attributes, docxDocumentInstance) => {
...
+      // the srtoke is solid by default
+      borderStrike = 'single';
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js]: move this into consts

File: src/helpers/xml-builder.js [hunk 15]
```diff
@@ -2030,105 +2353,151 @@ const buildTable = async (vNode, attributes, docxDocumentInstance) => {
...
+
+      setUpDirectionalBorderSize(tableBorders, borderSize);
+      tableBorders.strokes = setUpDirectionalBorderStroke(borderStrike);
+      tableBorders.colors = setUpDirectionalBorderColor(borderColor)
+
+      // in such cases, the inner cells also get a border of size 1
+      // these are not overwritten by the css border property of table tag
+      setUpDirectionalBorderSize(tableCellBorders, 1)
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js:2803]: move this into consts

File: src/helpers/xml-builder.js [hunk 16]
```diff
@@ -2030,105 +2353,151 @@ const buildTable = async (vNode, attributes, docxDocumentInstance) => {
...
+    }
+
+    if (Object.keys(tableStyles).length !== 0) {
+      for (const tableStyle of Object.keys(tableStyles)) {
+        if (tableStyle === 'border') {
+          const [cssSize, cssStroke, cssColor] = cssBorderParser(tableStyles.border, docxDocumentInstance.tableBorders);
+          borderSize = cssSize ?? borderSize;
+          borderColor = cssColor || borderColor;
+          borderStrike = cssStroke || borderStrike;
+          // TODO: Remove these comments when handling left and right borders
```
Comment (nicolasiscoding) [inline src/helpers/xml-builder.js:2816]: putting a note here so we don't forget

**Review by nicolasiscoding — CHANGES_REQUESTED** [review]
Code review complete, did not test for functionality but will compare it to the smoke test when ready

File: src/helpers/xml-builder.js (hunk 1 above)
Comment (K-Kumar-01) [reply src/helpers/xml-builder.js]: So if you provide hidden to a table border or table cell border then no matter what other properties are given, the borders will remain hidden.
Example: If you set for table, border-style as hidden and then to a table cell provide border-style as solid, no border will be shown for outer cells(i.e. no table border)

Now, in code we needed to distinguish it because if we provide none instead of hidden then the values are overwritten.

File: src/helpers/xml-builder.js (hunk 10 above)
Comment (K-Kumar-01) [reply src/helpers/xml-builder.js:1495]: Hidden is a special case. If given, no borders are shown no matter the priority. Thats why separately written

**Comment by nicolasiscoding** [comment]
Making ticket internally @K-Kumar-01
